lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp

```diff
@@ -819,27 +841,15 @@ SILCombiner::propagateConcreteTypeOfInitExistential(FullApplySite Apply,
   // are stuck:
   // We will re-create the same instruction and re-populate the worklist
   // with it.
-  if (CEI.ConcreteType != WMI->getLookupType()
-      || SelfConformance != WMI->getConformance()) {
-    SILBuilderWithScope WMIBuilder(WMI, BuilderCtx);
-    // Keep around the dependence on the open instruction unless we've
-    // actually eliminated the use.
-    auto *NewWMI = WMIBuilder.createWitnessMethod(
-        WMI->getLoc(), CEI.ConcreteType, SelfConformance, WMI->getMember(),
-        WMI->getType());
-    // Replace only uses of the witness_method in the apply that was analyzed by
-    // ConcreteExistentialInfo.
-    MutableArrayRef<Operand> Operands =
-        Apply.getInstruction()->getAllOperands();
-    for (auto &Op : Operands) {
-      if (Op.get() == WMI)
-        Op.set(NewWMI);
-    }
-    if (WMI->use_empty())
-      eraseInstFromFunction(*WMI);
+  if (CEI.ConcreteType != WMI->getLookupType() ||
+      SelfConformance != WMI->getConformance()) {
+    replaceWitnessMethodInst(Apply, WMI, BuilderCtx, CEI.ConcreteType,
+                             SelfConformance);
+    newWMIAdded = true;
```

Comment:
This is bizarre. It's fine to return a `bool`, but the convention in SILCombine, and the optimizer in general is that returning `true` indicates success. This function should return `true` whenever `createApplyWithConcreteType` is called.

If the caller needs to know whether the `witness_method` was _also_ needed to be rewritten then you have plenty of options:
- Check the `witness_method` lookup type in the caller.
- Check if the `witness_method` instruction has changed in the caller.
- Pass in a reference to a WitnessMethodInst* result.
- Define a more complicated return type `struct { bool success, WitnessMethodInst *newWMI }`

Reply:
I see. Returning bool was one of the suggestions made by @slavapestov as opposed to my code checking for WMI change in the caller, since we do not use the return type -- which made sense to me.

Reply:
Well, any of the above options for checking WMI are fine, but giving `propagateConcreteTypeOfInitExistential` a counterintuitive return value for the sake of convenience would be bad.